lib/construct.ex

```diff
@@ -61,17 +62,19 @@ defmodule Construct do
     quote do
       import Construct
 
+      Construct.__register_as_complex_type__(__MODULE__)
+
+      Module.register_attribute(__MODULE__, :required, accumulate: true)
+      Module.register_attribute(__MODULE__, :optional, accumulate: true)
```

Comment:
you don't need separate attributes for optional and required
if you really want `optional/2,3` macro — make an alias for `field/2,3` with options:

```elixir
defmacro optional(name, type, opts) do
  quote do
    Construct.__field__(__MODULE__, unquote(name), unquote(type), Keyword.put(unquote(opts), :default, nil]))
  end
end
```

I think that `required` macro is useless